### PR TITLE
Installers' custom data

### PR DESCRIPTION
**Pull request by arcanis**
**What's the problem this PR addresses?**

PR #2037 was storing custom data per-package, which led to a more complex code than needed since we needed to hydrate/persist the metadata before/after each `installPackage` call, and prevented to define custom data whose value was tied to the overall installer state.

**How did you fix it?**

Refactors to make the custom data part of each installer rather than individual per-package.

**Checklist**

- [x] I have read the [Contributing Guide](https://yarnpkg.com/advanced/contributing).

- [x] I have set the packages that need to be released for my changes to be effective.

- [x] I will check that all automated PR checks pass before the PR gets reviewed.

### CODE REVIEW

File: packages/plugin-node-modules/sources/NodeModulesLinker.ts [hunk 1]
```diff
@@ -105,30 +201,88 @@ class NodeModulesInstaller extends AbstractPnpInstaller {
...
       baseFs: defaultFsLayer,
       project: this.opts.project,
       report: this.opts.report,
-      loadManifest: this.cachedManifestLoad.bind(this, pnp, defaultFsLayer),
+      loadManifest: async locatorKey => {
+        const locator = structUtils.parseLocator(locatorKey);
+
+        const slot = this.localStore.get(locator.locatorHash);
+        if (typeof slot === `undefined`)
+          throw new Error(`Assertion failedd: Expected the slot to exist`);
```
Comment (larixer) [inline packages/plugin-node-modules/sources/NodeModulesLinker.ts:282]:
```suggestion
          throw new Error(`Assertion failed: Expected the slot to exist`);
```

File: packages/plugin-node-modules/sources/NodeModulesLinker.ts [hunk 2]
```diff
@@ -138,85 +292,47 @@ class NodeModulesInstaller extends AbstractPnpInstaller {
...
-
-      const dependencyMeta = this.opts.project.getDependencyMeta(locator, manifest.version);
-
-      if (buildScripts.length > 0 && dependencyMeta && dependencyMeta.built === false) {
-        this.opts.report.reportInfoOnce(MessageName.BUILD_DISABLED, `${structUtils.prettyLocator(this.opts.project.configuration, locator)} lists build scripts, but its build has been explicitly disabled through configuration.`);
-        buildScripts.length = 0;
-      }
+      const slot = this.localStore.get(locator.locatorHash);
+      if (typeof slot === `undefined`)
+        throw new Error(`Assertion faile: Expected the slot to exist`);
```
Comment (larixer) [inline packages/plugin-node-modules/sources/NodeModulesLinker.ts:297]:
```suggestion
        throw new Error(`Assertion failed: Expected the slot to exist`);
```

File: .github/workflows/integration-workflow.yml [hunk 3]
```diff
@@ -2,6 +2,7 @@ on:
   push:
     branches:
     - master
+    - mael/linker-refactoring
```
Comment (larixer) [inline .github/workflows/integration-workflow.yml:5]: Should it be there?

Comment (arcanis) [reply .github/workflows/integration-workflow.yml:5]: Nope, will remove it, I just worked on a separate branch where I needed tests to run.

File: packages/plugin-pnp/sources/PnpLinker.ts [hunk 4]
```diff
@@ -1,13 +1,12 @@
...
+import {FetchResult, Locator, Package}                                                                       from '@yarnpkg/core';
+import {Linker, LinkOptions, MinimalLinkOptions, Manifest, MessageName, DependencyMeta, LinkType, Installer} from '@yarnpkg/core';
+import {CwdFS, FakeFS, PortablePath, npath, ppath, xfs, Filename}                                            from '@yarnpkg/fslib';
+import {generateInlinedScript, generateSplitScript, PackageRegistry, PnpSettings}                            from '@yarnpkg/pnp';
+import {UsageError}                                                                                          from 'clipanion';
 
-import {AbstractPnpInstaller}                                                           from './AbstractPnpInstaller';
-import {getPnpPath}                                                                     from './index';
-import * as pnpUtils                                                                    from './pnpUtils';
+import {getPnpPath, javascriptUtils}                                                                         from './index';
```
Comment (merceyz) [inline packages/plugin-pnp/sources/PnpLinker.ts:8]: Impressive this doesn't/didn't throw at runtime because this seems like a circular import

**Comment by arcanis** [comment]
Will merge now and see over time what things look like on the benchmark. A manual run seems to be *slower*, but it's difficult to know if that's just a fluke or something more serious (local runs are definitely faster). Worst case we can revert later on.

**Pull request merged**